### ISSUE TITLE
`int | str` is not a valid `type[int | str]`

### ISSUE THREAD

**Issue by KotlinIsland** [microsoft/pyright#2522]
```py
a: type[int | str] = int | str  # no error
```
PyCharm correctly handles this.

**Comment by erictraut** [comment]
What is the issue? Do you think this should generate an error? Why?

Python 3.10 accepts it at runtime. It's valid syntax with unambiguous semantic meaning, and it's type consistent. Mypy doesn't generate an error.

Are you suggesting that a type checker should instead force you to use `a: type[int] | type[str] = int | str`? Or do you think that should also generate an error?

This does become a problem at runtime if you attempt to instantiate `a` by calling its constructor: `a()`, but that's true of any union of types.

**Comment by KotlinIsland** [comment]
Mypy shows an error with this.

`int | str` is not a valid `type[int]` and it is not a valid `type[str]`. It can never match any of the bounds of that annotation, it's an instance of `UnionType`.

**Comment by erictraut** [comment]
OK, I do see an error with the latest version of mypy.

I'll look into adding an error for this case.

**Comment by KotlinIsland** [comment]
> This does become a problem at runtime if you attempt to instantiate a by calling its constructor: a(), but that's true of any union of types.

Do you mean that all union types are incorrectly implemented in pyright? I don't understand what you mean.

**Comment by KotlinIsland** [comment]
eg:
```py
class Foo:
    a: int
class Bar:
    a: int

a: type[Foo] | type[Bar] = Foo | Bar
print(a.a)
```
`UnionType` has no member `a`.

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.186, which I just published. It will also be included in the next release of pylance.
